examples/api/lib/material/context_menu/context_menu_controller.0.dart

```diff
@@ -0,0 +1,154 @@
+            // The custom context menu will look like the default context menu
+            // on the current platform, but it will have only a single 'Print'
+            // button.
+            return DefaultTextSelectionToolbar(
+              primaryAnchor: primaryAnchor,
+              secondaryAnchor: secondaryAnchor,
+              buttonItems: <ContextMenuButtonItem>[
+                ContextMenuButtonItem(
+                  onPressed: () {
+                    ContextMenuController.hide();
+                    _showDialog(context);
+                  },
+                  label: 'Print',
+                ),
+              ],
+            );
+          },
+          // In this case we're wrapping a big open space in a GestureDetector
+          // in order to show the context menu, but we could also wrap a single
+          // wiget like an Image to give it a context menu.
+          child: ListView(
+            children: <Widget>[
+              Container(height: 20.0),
+              const Text('Right click or long press anywhere to show the custom menu.'),
+            ],
+          ),
+        ),
+      ),
+    );
+  }
+}
+
+/// Shows and hides the context menu based on user gestures.
+///
+/// By default, shows the menu on right clicks and long presses.
+class _ContextMenuRegion extends StatefulWidget {
+  /// Creates an instance of [_ContextMenuRegion].
+  const _ContextMenuRegion({
+    required this.child,
+    required this.contextMenuBuilder,
+  });
+
+  /// Builds the context menu.
+  final ContextMenuBuilder contextMenuBuilder;
+
+  /// The child widget that will be listened to for gestures.
+  final Widget child;
+
+  @override
+  State<_ContextMenuRegion> createState() => _ContextMenuRegionState();
+}
+
+class _ContextMenuRegionState extends State<_ContextMenuRegion> {
+  Offset? _longPressOffset;
+
+  static bool get _longPressEnabled {
+    switch (defaultTargetPlatform) {
+      case TargetPlatform.android:
+      case TargetPlatform.iOS:
+        return true;
+      case TargetPlatform.macOS:
+      case TargetPlatform.fuchsia:
+      case TargetPlatform.linux:
+      case TargetPlatform.windows:
+        return false;
+    }
+  }
+
+  void _onSecondaryTapUp(TapUpDetails details) {
+    _show(details.globalPosition);
+  }
+
+  void _onTap() {
+    if (!ContextMenuController.isShown) {
+      return;
+    }
+    _hide();
+  }
+
+  void _onLongPressStart(LongPressStartDetails details) {
+    _longPressOffset = details.globalPosition;
+  }
+
+  void _onLongPress() {
+    assert(_longPressOffset != null);
+    _show(_longPressOffset!);
+    _longPressOffset = null;
+  }
+
+  void _show(Offset position) {
+    ContextMenuController.show(
+      context: context,
+      contextMenuBuilder: (BuildContext context) {
+        return widget.contextMenuBuilder(context, position);
+      },
+    );
+  }
+
+  void _hide() {
+    ContextMenuController.hide();
```

Comment:
I find it easy to make mistakes in apps where more than one thing can present a context menu: there doesn't seem to be a way to know whose menu is currently being presented and when you call `hide` (especially in `dispose`), chances are you're hiding the context menu that another widget just presented.

Reply:
I think I'm going to take a closer look at ContextMenuController.  The whole idea of having this be global scares me, I was hoping someone would tell me it's totally forbidden in Flutter or something.  It's also true that `isShown` doesn't necessarily tell you that *your* menu is shown.  Maybe I can solve all of these problems by allowing this to be instantiated and refactoring a bit somehow.

Reply:
I've refactored ContextMenuController to not be a singleton, see https://github.com/flutter/flutter/pull/107193#issuecomment-1241333434